### PR TITLE
Adding Using the `Text Generation Inference (TGI)` with Llama Models

### LINKED ISSUE #45: Serve Llama models with `Text Generation Inference` recipe
This issue comes from the open call for contributions #43.

In this case, this issue is intended for the generation of a recipe for serving Llama models using the [Text Generation Inference](https://huggingface.co/docs/text-generation-inference/en/index)

### PR DESCRIPTION
**Pull request by sergiopaniego**
Adding a TGI recipe that includes examples of how to serve and consume the 🦙 Llama models using different methods available in the HF ecosystem. 

fix #45

@ariG23498

### CODE REVIEW

**Comment by sergiopaniego** [comment]
I'm not sure if the file location is correct, and we might need to link the recipe in the README or somewhere else, so that may need updating as well. 😄

**Comment by ariG23498** [comment]
@sergiopaniego I really like the notebook! :fire: 

Let's have the current notebook broken down into two parts:
1. A shell script to run TGI with llama model using docker
2. A separate notebook that holds the consuming the endpoint recipe? I really like the chat with image bit! (awesome work)

CC: @Vaibhavs10 what do you think?

**Comment by sergiopaniego** [comment]
Thanks a lot for the feedback, @ariG23498!

I think it makes sense to split the notebook into two separate parts, as you suggested. Should I start working on it now, or do we need a double-check from @Vaibhavs10? 😄

**Review by pcuenca — COMMENTED** [review]
Actually, I think having everything self-contained inside a notebook is nice because we can provide explanations and guidelines. A script with comments would be more difficult to understand in my opinion.

I have a few suggestions. You don't have to address everything, they are just meant to improve the reading experience as I'm going through the notebook:

- Let's put this inside a directory, and link from the main README.
- The beginning feels heavy with section headers. I'd remove the headings for `Available Llama Models` and `Learn More About TGI` but include the same content as sentences/paragraphs.
- There's no need to bold the section titles.
- In the `Using Docker` section, make it clear these are command-line commands to be run in the terminal. Also, I don't know if it's intentional but they are not code snippets.
- We can use something like `TOKEN=$(cat ~/.cache/huggingface/token)` in various places to avoid the use of placeholders.
- Related: we should maybe remind people to authenticate (`huggingface-cli login`)  using a token from [here](https://huggingface.co/settings/tokens)
- In the Inference Endpoints section, I'd mention the availability of llama.cpp deployments for GGUF quantized models. This way people may be able to use smaller instances.
- Is there a reason we are using a Vision model for text-only queries? It's fine that we do, but maybe explain that these models don't require images to work in text-only mode so people are not puzzled.
- For the curl chat completion example, if we are using streaming mode it would be awesome to have a simple awk script to process the outputs and print just the content.
- Using Python: typo `mode` -> `more`
- Remove the comments below the `client` lines in the `InferenceClient` examples. They mention a `base_model` parameter while the actual code lines use `model`.
- InferenceClient chat example: write in the same line instead to avoid a huge output box.
- Chat with Image: `The Llama models` -> `The Llama Vision models`.
- We don't really need matplotlib to display the image, do we? `img` as the last line of the cell should just work :)
- I'd maybe try to ask some concrete questions about the image, I think the result would be more impressive :) It doesn't have to be a technical one either.
- For the last example, if you put the `{"type": "text", "text": "What’s in this image?"}` **below** the `image_url` you may get better results. From my tests, it seems that the model was primarily trained with leading images.

(Sorry for not providing an inline review, it's a bit difficult to review notebooks, unfortunately).

**Review by Vaibhavs10 — COMMENTED** [review]
Sorry for the delay in responding - I pretty much +1 all that @pcuenca mentioned 🤗

**Comment by sergiopaniego** [comment]
Thanks a lot for the in-depth review! 

I'll update the PR accordingly 👨‍💻

**Comment by sergiopaniego** [comment]
I've updated the notebook covering all the possible improvements mentioned @pcuenca. 🤗 

Open to any further suggestions or enhancements!

cc @ariG23498 @Vaibhavs10

**Review by Vaibhavs10 — COMMENTED** [review]
Very nice! Great job on this! 🔥

One quick point: It'd be great to clarify that all of these models are deployed via TGI on the serverless API too (that's the same thing you're calling with `huggingface_hub` client)

Another thing is that it says TGI recipe but we only cover a small part of TGI and Rest is showing inference via Inference API maybe we can rename the folder more appropriately?

File: .github/README.md [hunk 1]
```diff
@@ -150,4 +150,10 @@ on the rise. Here we show you how to build your very own synthetic dataset.
 ## Llama RAG 
 Seeking an entry-level RAG pipeline? This notebook guides you through building a very simple streamlined RAG experiment using Llama and Hugging Face.
 
-* [Simple RAG Pipeline](../llama_rag/llama_rag_pipeline.ipynb)
\ No newline at end of file
+* [Simple RAG Pipeline](../llama_rag/llama_rag_pipeline.ipynb)
+
+
+## Text Generation Inference (TGI) with Llama Models
```
Comment (Vaibhavs10) [inline .github/README.md:156]:
```suggestion
## Text Generation Inference (TGI) & API Inference with Llama Models
```

File: .github/README.md [hunk 2]
```diff
@@ -150,4 +150,10 @@ on the rise. Here we show you how to build your very own synthetic dataset.
...
+In addition to efficient inference with large models, the Text Generation Inference (TGI) framework enables scalable deployment of Llama models. Learn how to integrate TGI for fast text generation in the following recipe:
```
Comment (Vaibhavs10) [inline .github/README.md:157]:
```suggestion
Text Generation Inference (TGI) framework enables efficient and  scalable deployment of Llama models. In this notebook we'll learn how to integrate TGI for fast text generation and to consume already deployed Llama models via Inference API:
```

**Review by Vaibhavs10 — APPROVED** [review]
Left two minor suggestions for wording.

**Comment by sergiopaniego** [comment]
Added @Vaibhavs10 😄! Thanks a lot for all the comments and suggestions!!

**Comment by Vaibhavs10** [comment]
Perfect from myside to merge - I'd let @pcuenca take the final pass and merge! 🤗

File: .github/README.md [hunk 3]
```diff
@@ -150,4 +150,10 @@ on the rise. Here we show you how to build your very own synthetic dataset.
...
 
-* [Simple RAG Pipeline](../llama_rag/llama_rag_pipeline.ipynb)
\ No newline at end of file
+* [Simple RAG Pipeline](../llama_rag/llama_rag_pipeline.ipynb)
+
+
+## Text Generation Inference (TGI) & API Inference with Llama Models
+Text Generation Inference (TGI) framework enables efficient and  scalable deployment of Llama models. In this notebook we'll learn how to integrate TGI for fast text generation and to consume already deployed Llama models via Inference API:
+
+* [Text Generation Inference (TGI) with Llama Models](../llama_tgi/tgi_recipe.ipynb)
```
Comment (pcuenca) [inline .github/README.md:159]:
```suggestion
* [Text Generation Inference (TGI) with Llama Models](../llama_tgi_api_inference/tgi_api_inference_recipe.ipynb) 
```

**Review by pcuenca — APPROVED** [review]
Nits on the notebook:

* Maybe mention access permissions as well as gating
`These models are gated, so authentication is required before accessing them. You also need to have been granted access to the models; if you still haven't requested it, it's enough to request it for [one of the models in the collection](https://huggingface.co/meta-llama/Llama-3.2-1B-Instruct).`

* Earlier mention so the transition from TGI Docker to Inference Endpoints does not look like a change of course:
`Another efficient way to serve Llama models is by using **Inference Endpoints** on Hugging Face, which uses TGI under the hood.`

* Just for info, in Python you can do `huggingface_hub.get_token()` to retrieve the logged-in token. However, it's nice to do it without the dependency too.

* Another earlier mention in the "Using cURL" section for maybe better guidance:
`To generate text using the Llama model, you can send a POST request like follows. For these examples we are using [free, pre-deployed endpoints](https://huggingface.co/docs/api-inference/en/index), but you can change the URL to refer to your own TGI or Inference Endpoints deployment.`
(It's fine that we repeat something similar later)

**Comment by sergiopaniego** [comment]
Thanks for the review and suggestions @pcuenca!
I've improved the notebook based on them 😄

**Review by Vaibhavs10 — APPROVED** [review]

**Pull request merged**